### PR TITLE
Propagate zeek-version.h skip via CMake properties

### LINKED ISSUE #2806: Builtin external plugins and zeek/zeek-version.h backwards compat
With #2802, ZEEK_VERSION* macros were split from zeek-config.h into zeek-version.h. For backwards compatibility, when building an external plugin, zeek-config.h does include zeek-version.h. This mechanism does not extend to "builtin external" plugins - those configured with `--include-plugins`. If they reference ZEEK_VERSION_NUMBER, zeek/zeek-version.h must be included explicitly. For AF_Packet and spicy-plugin that was done in zeek/zeek-af_packet-plugin#45, zeek/spicy-plugin#48.

If we could string through a `-DZEEK_BUILTIN_EXTERNAL_PLUGIN=true` or so, that would allow extending the `zeek/zeek-version.h` include condition also to builtin plugins. I don't see how we otherwise know we're building an external plugin.

For most users, there's no impact, but minimally @bbannier was wondering if that isn't too fragile. Other opinions?

Maybe it's not worth the trouble, in which case we can close this again.

This is where builtin external plugins are added, I'd guess we could push and pop to the compile options, but @Neverlord may have better ideas: https://github.com/zeek/zeek/blob/ebc956324358d194c77dd2fe67572fd10ca35b9f/src/CMakeLists.txt#L214-L222

### PR DESCRIPTION
**Pull request by Neverlord**
Add a define for `ZEEK_CONFIG_SKIP_VERSION_H` to static plugins as well as to dynamic plugins that we build alongside Zeek. When including `zeek-config.h` with this macro defined, the header skips including `zeek-version.h`.

For static plugins, we always define `ZEEK_CONFIG_SKIP_VERSION_H`.

For dynamic plugins, we define `ZEEK_CONFIG_SKIP_VERSION_H` on the build interface via the `Zeek::DynamicPluginBase` target. The build interface includes everything that gets pulled in from our main CMake file, but also plugins that build against the source tree of Zeek (via `ZEEK_DIST`) by reading `build/ZeekTargets.cmake`:

```
...
set_target_properties(Zeek::DynamicPluginBase PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "ZEEK_CONFIG_SKIP_VERSION_H;DEBUG")
...
```

On the installed `<prefix>/share/zeek/cmake/ZeekTargets.cmake`, this define is missing:

```
...
set_target_properties(Zeek::DynamicPluginBase PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "DEBUG"
...
```

@awelzel I think that's the semantics you've been looking for, right? Please tell me if I got something wrong here. If plugins that build against the source tree via `ZEEK_DIST` should *not* skip the header, we need to set the define in `ZeekPluginDynamic.cmake`. 🙂

Closes #2806.

### CODE REVIEW

**Comment by awelzel** [comment]
Thanks taking this on @Neverlord. 

I don't think it's quite right, currently. The main twist is that this is less about static vs dynamic and more about in-tree vs out-of-tree plugins.

* For "in-tree builtin plugins" (logging/writeres/ascii), zeek-version.h should be skipped to avoid the ccache busting of in-tree plugins. In-tree source code never needs ZEEK_VERSION_NUMBER conditionals and therefore zeek-version.h isn't needed.

* For "out-of-tree builtin plugins" (those pulled in via ./configure --include-plugins=...), zeek-version.h has to be included. These are external/out-of-tree and so may have valid ZEEK_VERSION_NUMBER conditionals. These are compiled into Zeek statically though. They are "added" here:
https://github.com/zeek/zeek/blob/0b22f792a56719e52ba63b9d4f7260fee633a67d/src/CMakeLists.txt#L205-L213

* For "dynamic plugins", can currently always consider them out-of tree and never skip the zeek-version.h include.

> @awelzel I think that's the semantics you've been looking for, right? Please tell me if I got something wrong here. If plugins that build against the source tree via ZEEK_DIST should not skip the header, we need to set the define in ZeekPluginDynamic.cmake. slightly_smiling_face

Do above examples help? zeek-version.h shouldn't be skipped when ZEEK_DIST as that implies we're compiling an external plugin. Sample use-case would be checking out zeek-master and zeek-5.1 in separate work-trees and testing ZEEK_VERSION_NUMBER within the plugin's code.
```
./configure --zeek-dist=/path/to/zeek-master

./configure --zeek-dist=/path/to/release-5.1
```

Hope this help. Message me if anything is left unclear.

**Comment by Neverlord** [comment]
@awelzel next try. This sets the SKIP flag only for static plugins that are *not* pulled in from `BUILTIN_PLUGIN_LIST` (that gets populated from the `configure` script and also contains spicy-plugin and ZeekJS).

**Comment by awelzel** [comment]
> @awelzel next try. This sets the SKIP flag only for static plugins that are not pulled in from BUILTIN_PLUGIN_LIST (that gets populated from the configure script and also contains spicy-plugin and ZeekJS).

Thanks - I've left a naming thought in the cmake/ PR, no strong opinion and that seems to work :-) 

Would you consider picking up the following commit - possibly change to fit your style? https://github.com/zeek/zeek/commit/69ca942774398d0653b6ece9019b97f540551523

It adds a stub builtin plugin into the builtin-plugins testing directory and hooks it up in cirrus (I hope I got that right). With current master this currently fails to compile and after your fix it should work because ZEEK_VERSION_NUMBER comes in through zee/zeek-config.h then.

Seems to work here, but haven't tried in CI.

**Comment by awelzel** [comment]
> require that they can do nasty stuff like adding extra dependencies directly into zeekdeps. Like the JS plugin here, but there are more examples.

Mostly as reference, below is the snippet from ZeekJS (here's [spicy-plugin's version](https://github.com/zeek/spicy-plugin/blob/cf1cf6baca25964a8cc29bc004755d02a6aebd09/CMakeLists.txt#L168-L172). These were added when `--include-plugins` usage didn't properly propagate libraries through `zeek_plugin_link_library()`. That has been fixed now.

```
if ( ZEEK_PLUGIN_INTERNAL_BUILD )
    # This is a hack for the current bug in zeek with doing built in plugins
    set(zeekdeps ${zeekdeps} ${NODEJS_LIBRARIES} PARENT_SCOPE)
else()
    zeek_plugin_link_library("${NODEJS_LIBRARIES}")
endif()
```

> Just wanted to raise awareness for this a bit. The idea was to shift more and more towards a cleaner, target-based approach with CMake. Having plugins force us to operate on global variables is definitely holding that back.

I could see breaking the above to have the nicer scoping as you propose. While seemingly a regular topic in issues/tickets, external-static plugins are advanced territory for advanced users. They come up because they had and continue to have quirks.

So from my side, if the `zeekdeps` hack doesn't work with 6.0, but `zeek_plugin_link_library()` does reliably, that would be okay.

File: zeek-config.h.in [hunk 1]
```diff
@@ -248,19 +248,10 @@
 /* String with extension of dynamic libraries (e.g., ".so") */
 #define DYNAMIC_PLUGIN_SUFFIX "@CMAKE_SHARED_MODULE_SUFFIX@"
 
-/* Set if we're building outside of the main Zeek source code tree. */
-#ifndef ZEEK_PLUGIN_INTERNAL_BUILD
-#define ZEEK_PLUGIN_INTERNAL_BUILD true
```
Comment (awelzel) [inline zeek-config.h.in:253]: I was forth and back to say we should keep this, but I found that we removed the define `ZEEK_PLUGIN_INTERNAL_BUILD=false` for external plugins during the re-work already: https://github.com/zeek/cmake/commit/a9146d635357ef61cbf0e637df6d678d28e4a702#diff-7d1e0599dcdb8f6a71c933643b3d90adb768c608f62dc87067d847dd9b703036L242

So this is probably behaving wrongly with the current state for external plugins anyhow (?) and if the fix is to removes the define, it's a :+1: from me.

**Review by awelzel — APPROVED** [review]
Thanks for leaving the notes around macro vs function, I'll move it in as is.

**Pull request merged**